src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java

```diff
@@ -0,0 +1,365 @@
+                } else {
+                    LOGGER.log(WARNING, "could not cancel an older flow because it has no assigned executor");
+                }
+            }
+        }
+    }
+
+    private static void println(StepContext context, String message) {
+        if (!context.isReady()) {
+            LOGGER.log(Level.FINE, "cannot print message ‘{0}’ to dead {1}", new Object[] {message, context});
+            return;
+        }
+        try {
+            context.get(TaskListener.class).getLogger().println(message);
+        } catch (Exception x) {
+            LOGGER.log(WARNING, "failed to print message to dead " + context, x);
+        }
+    }
+
+    private static void cancel(StepContext context, Integer build) throws IOException, InterruptedException {
+        if (context.isReady()) {
+            println(context, "Canceled since build #" + build + " already got here");
+            Run<?, ?> r = context.get(Run.class);
+            String job = "";
+            if (r != null) { // it should be always non-null at this point, but let's do a defensive check
+                job = r.getParent().getFullName();
+            }
+            throw new FlowInterruptedException(Result.NOT_BUILT, new CancelledCause(job + "#" + build));
+        } else {
+            LOGGER.log(WARNING, "cannot cancel dead #" + build);
+        }
+    }
+
+    private static void cleanUp(Job<?,?> job, String jobName) {
+        Map<Integer, Milestone> milestonesInJob = getMilestonesByOrdinalByJob().get(jobName);
+        assert milestonesInJob != null;
+        Iterator<Entry<Integer, Milestone>> it = milestonesInJob.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<Integer, Milestone> entry = it.next();
+            Set<Integer> inSight = entry.getValue().inSight;
+            Iterator<Integer> it2 = inSight.iterator();
+            while (it2.hasNext()) {
+                Integer number = it2.next();
+                if (job.getBuildByNumber(number) == null) {
+                    // Deleted at some point but did not properly clean up from exit(…).
+                    LOGGER.log(WARNING, "Cleaning up apparently deleted {0}#{1}", new Object[] {jobName, number});
+                    it2.remove();
+                }
+            }
+        }
+    }
+
+    private static void load() {
+        Jenkins.getActiveInstance().getDescriptorOrDie(MilestoneStep.class).load();
+    }
+
+    private static void save() {
+        Jenkins.getActiveInstance().getDescriptorOrDie(MilestoneStep.class).save();
+    }
+
+    @Extension
+    public static final class Listener extends RunListener<Run<?,?>> {
+        @Override public void onCompleted(Run<?,?> r, TaskListener listener) {
+            if (!(r instanceof FlowExecutionOwner.Executable) || ((FlowExecutionOwner.Executable) r).asFlowExecutionOwner() == null
+                    || Result.ABORTED.equals(r.getResult())) {
```

Comment:
…though even under the first assumption, you cannot _reliably_ distinguish based solely on `Result` whether a build ran to the “end” of the script, you can only guess. A script can `return` at any point without affecting result; and it can set `currentBuild.result` directly to some value, rather than relying on it being set implicitly by Pipeline infrastructure based on a `Throwable`. For that matter, an uncaught `FlowInterruptedException` can encode any `Result`, not just `ABORTED`.